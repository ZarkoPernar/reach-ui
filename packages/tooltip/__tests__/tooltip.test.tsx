import React from "react";
import { render, fireEvent, act } from "$test/utils";
import Tooltip, { LEAVE_TIMEOUT, MOUSE_REST_TIMEOUT } from "@reach/tooltip";

const { keyDown, mouseOver, mouseLeave, focus, blur } = fireEvent;

describe("<Tooltip />", () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  it("shows/hides on hover", () => {
    const { baseElement, getByText } = render(
      <Tooltip label="Content">
        <button>Trigger</button>
      </Tooltip>
    );
    const trigger = getByText("Trigger");

    expect(baseElement).toMatchSnapshot("not visible");

    act(() => void mouseOver(trigger));
    act(() => void jest.advanceTimersByTime(MOUSE_REST_TIMEOUT));
    expect(baseElement).toMatchSnapshot("after mouse rest timeout");

    act(() => void mouseLeave(trigger));
    act(() => void jest.advanceTimersByTime(LEAVE_TIMEOUT));
    expect(baseElement).toMatchSnapshot("after leave timeout");
  });

  it("renders as any HTML element", () => {
    const { getByText } = render(
      <p>
        <Tooltip as="span" style={{ display: "block" }} label="Content">
          <span>Trigger</span>
        </Tooltip>
      </p>
    );

    const trigger = getByText("Trigger");
    act(() => void mouseOver(trigger));
    act(() => void jest.advanceTimersByTime(MOUSE_REST_TIMEOUT));
    const tooltip = getByText("Content");
    expect(tooltip.tagName).toBe("SPAN");
  });

  it("shows/hides when trigger is activeElement", () => {
    const { baseElement, getByText } = render(
      <Tooltip label="Content">
        <button>Trigger</button>
      </Tooltip>
    );
    const trigger = getByText("Trigger");

    expect(baseElement).toMatchSnapshot("not visible");

    act(() => void focus(trigger));
    expect(baseElement).toMatchSnapshot("after focus");

    act(() => void blur(trigger));
    act(() => void jest.advanceTimersByTime(LEAVE_TIMEOUT));
    expect(baseElement).toMatchSnapshot("after blur");
  });

  it("shows without timeout when one tooltip is already visible", () => {
    const { baseElement, getByText } = render(
      <>
        <Tooltip label="First">
          <button>First Trigger</button>
        </Tooltip>
        <Tooltip label="Second">
          <button>Second Trigger</button>
        </Tooltip>
      </>
    );
    const firstTrigger = getByText("First Trigger");
    const secondTrigger = getByText("Second Trigger");

    act(() => void mouseOver(firstTrigger));
    act(() => void jest.advanceTimersByTime(MOUSE_REST_TIMEOUT));
    expect(baseElement).toMatchSnapshot("after mouse rest timeout");

    act(() => void mouseLeave(firstTrigger));
    act(() => void mouseOver(secondTrigger));
    expect(baseElement).toMatchSnapshot("after switch without timeout");
  });

  it("hides on ESC", () => {
    const { baseElement, getByText } = render(
      <Tooltip label="Content">
        <button>Trigger</button>
      </Tooltip>
    );

    const trigger = getByText("Trigger");

    focus(trigger);
    act(() => void jest.advanceTimersByTime(MOUSE_REST_TIMEOUT));
    expect(baseElement).toMatchSnapshot("after mouse rest timeout");

    act(() => void keyDown(trigger, { key: "Escape" }));
    expect(baseElement).toMatchSnapshot("not visible after ESC");
  });
});
