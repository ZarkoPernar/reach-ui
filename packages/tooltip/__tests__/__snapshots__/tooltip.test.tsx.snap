// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tooltip /> hides on ESC: after mouse rest timeout 1`] = `
<body>
  <div>
    <button
      aria-describedby="tooltip--REACH_ID_"
      data-reach-tooltip-trigger=""
    >
      Trigger
    </button>
  </div>
  <reach-portal>
    <div
      data-reach-tooltip="true"
      id="tooltip--REACH_ID_"
      role="tooltip"
      style="left: 0px; top: 8px;"
    >
      Content
    </div>
  </reach-portal>
</body>
`;

exports[`<Tooltip /> hides on ESC: not visible after ESC 1`] = `
<body>
  <div>
    <button
      data-reach-tooltip-trigger=""
    >
      Trigger
    </button>
  </div>
</body>
`;

exports[`<Tooltip /> shows without timeout when one tooltip is already visible: after mouse rest timeout 1`] = `
<body>
  <div>
    <button
      aria-describedby="tooltip--REACH_ID_"
      data-reach-tooltip-trigger=""
    >
      First Trigger
    </button>
    <button
      aria-describedby="tooltip--REACH_ID_"
      data-reach-tooltip-trigger=""
    >
      Second Trigger
    </button>
  </div>
  <reach-portal>
    <div
      data-reach-tooltip="true"
      id="tooltip--REACH_ID_"
      role="tooltip"
      style="left: 0px; top: 8px;"
    >
      First
    </div>
  </reach-portal>
  <reach-portal>
    <div
      data-reach-tooltip="true"
      id="tooltip--REACH_ID_"
      role="tooltip"
      style="left: 0px; top: 8px;"
    >
      Second
    </div>
  </reach-portal>
</body>
`;

exports[`<Tooltip /> shows without timeout when one tooltip is already visible: after switch without timeout 1`] = `
<body>
  <div>
    <button
      aria-describedby="tooltip--REACH_ID_"
      data-reach-tooltip-trigger=""
    >
      First Trigger
    </button>
    <button
      aria-describedby="tooltip--REACH_ID_"
      data-reach-tooltip-trigger=""
    >
      Second Trigger
    </button>
  </div>
  <reach-portal>
    <div
      data-reach-tooltip="true"
      id="tooltip--REACH_ID_"
      role="tooltip"
      style="left: 0px; top: 8px;"
    >
      First
    </div>
  </reach-portal>
  <reach-portal>
    <div
      data-reach-tooltip="true"
      id="tooltip--REACH_ID_"
      role="tooltip"
      style="left: 0px; top: 8px;"
    >
      Second
    </div>
  </reach-portal>
</body>
`;

exports[`<Tooltip /> shows/hides on hover: after leave timeout 1`] = `
<body>
  <div>
    <button
      data-reach-tooltip-trigger=""
    >
      Trigger
    </button>
  </div>
</body>
`;

exports[`<Tooltip /> shows/hides on hover: after mouse rest timeout 1`] = `
<body>
  <div>
    <button
      aria-describedby="tooltip--REACH_ID_"
      data-reach-tooltip-trigger=""
    >
      Trigger
    </button>
  </div>
  <reach-portal>
    <div
      data-reach-tooltip="true"
      id="tooltip--REACH_ID_"
      role="tooltip"
      style="left: 0px; top: 8px;"
    >
      Content
    </div>
  </reach-portal>
</body>
`;

exports[`<Tooltip /> shows/hides on hover: not visible 1`] = `
<body>
  <div>
    <button
      data-reach-tooltip-trigger=""
    >
      Trigger
    </button>
  </div>
</body>
`;

exports[`<Tooltip /> shows/hides when trigger is activeElement: after blur 1`] = `
<body>
  <div>
    <button
      data-reach-tooltip-trigger=""
    >
      Trigger
    </button>
  </div>
</body>
`;

exports[`<Tooltip /> shows/hides when trigger is activeElement: after focus 1`] = `
<body>
  <div>
    <button
      data-reach-tooltip-trigger=""
    >
      Trigger
    </button>
  </div>
</body>
`;

exports[`<Tooltip /> shows/hides when trigger is activeElement: not visible 1`] = `
<body>
  <div>
    <button
      aria-describedby="tooltip--REACH_ID_"
      data-reach-tooltip-trigger=""
    >
      Trigger
    </button>
  </div>
  <reach-portal>
    <div
      data-reach-tooltip="true"
      id="tooltip--REACH_ID_"
      role="tooltip"
      style="left: 0px; top: 8px;"
    >
      Content
    </div>
  </reach-portal>
</body>
`;
